hazelcast/include/hazelcast/util/SynchronizedMap.h

```diff
@@ -69,10 +70,17 @@ namespace hazelcast {
              *         or <tt>null</tt> if there was no mapping for the key
              */
             std::shared_ptr<V> putIfAbsent(const K &key, std::shared_ptr<V> value) {
-                std::lock_guard<std::mutex> lg(mapLock);
-                if (internalMap.count(key) > 0) {
-                    return internalMap[key];
+                boost::upgrade_lock<boost::shared_mutex> read_lock(mapLock);
+                auto found = internalMap.find(key);
+                if (found != internalMap.end()) {
+                    return found->second;
                 } else {
+                    // upgrade to write lock
+                    boost::upgrade_to_unique_lock<boost::shared_mutex> write_lock(read_lock);
+                    found = internalMap.find(key);
```

Comment:
I think it is a right question to ask. Is "locking with unique_lock" is better in the first place or is it better with this way to search the key again after locking the write_lock? I think it may change according to the quality of hash function and key distrubition to the hash buckets or according to the number of threads, etc.